### ISSUE TITLE
Support for referencing other keys in translation definitions

### ISSUE THREAD

**Issue by Majkl578** [Kdyby/Translation#23]
Assume we have following translation definition file:

``` yml
homepage:
    foo: "Hello World"
    bar: "Hello Friend"
    baz: "Hello World"
```

The reasons may be as simple as different translations in different languages, but in some language, few translations would be same (e.g. czech vs. english).

It would be nice to avoid duplication of these messages, so something like this would do so:

``` yml
homepage:
    foo: "Hello World"
    bar: "Hello Friend"
    baz: %homepage.foo%
```

 And maybe referencing also other dictionary, distinguished by dot at the beginning:

``` yml
homepage:
    foo: "Hello World"
    bar: "Hello Friend"
    baz: %.differentDictionary.abc.foo%
```

Your idea? I may implement it afterwards.

**Comment by janedbal** [comment]
I've met similar issue when trying to accomplish following use-case:

```yaml
phone:
    default: 'Phone number %number% is not valid for state %state%.'
    prefixMismatch: %phone.default%' Country code %givenPrefix% belongs to %givenPrefixState%. Expected country code is %expectedPrefix%.'
    prefixInvalid: %phone.default%' Country code is not valid.'
    invalid: %phone.default%' Please check number of digits and typing errors.'
    invalidFormat: %phone.default%
    tooLong: %phone.default%' The number is too long.'
    tooShort: %phone.default%' The number is too short.'
```

With such concatenation, you wouldn't have to repeat the base error message and its parameters all over again.

**Comment by lookyman** [comment]
Well it might be better to just raise two separate errors with the two messages, and display both.

Though I quite like the idea behind this issue.

**Comment by fprochazka** [comment]
I really like @janedbal 's use-case, because he want's to replace "parameters" in the file, not other messages.

**Comment by MartinMystikJonas** [comment]
From my POW to be able to reference other massages can be useful. We usually create different messages for each context of message to by able later change it only on one place when needed. But then when you want to change it in all places it could require massive copy-paste. I think option to reference other message will improve DRY.

**Comment by fprochazka** [comment]
@MartinMystikJonas the question is... do you really wanna promote DRY in translation dictionaries? I highly disagree with that for reasons already explained. But, I'm not oposed to the general preprocessor idea.

**Comment by MartinMystikJonas** [comment]
@fprochazka Why not promote DRY?

I understand problem you describing but I don't agree with solution. Prom my POW this could prevent problem you describing. I agree that sometimes messages are same in different contexty only by acciedent not by design. And that change of one message then can result in non-sense in different context.

Bur real question is how to prevent this problem?

Problome with your solution is that people are layzy and I think thy will re-use existing messages rather than copy-pasting them. Then you have message referenced from different laces all across application.

From my experience it is better to have one key for each context and then use referencing inside translation files to remove need of copy-pasting message. When you then want to change message you can easily check if any other messages are referencing this message and if it is ok. It is easier when you need to check only translation files not whole application. Also this helps when change is intentional to all places (typo, misslelled word, ...).

So in summary: I think that possibility of referencing messages will promote creating new key for each context which what helps keep translation DRY and preven accidental modification in wrong context.

**Comment by Olicek** [comment]
I would need something similar, but conversaly. In company, where i works they use for translating static classes with inheritance. They use it successfully for large application with 30+ domains. It is legacy code, but i like idea of inheritence. For example:

```
class LNG_Main
{
  // Obecny obsah mailu
  public static $textStart = "Vážený zákazníku";
  public static $contactUs = "V případě jakýchkoli nejasností nás kontaktujte";
```

```
class LNG_Order extends LNG_Main
{
  // Obecne texty
  public static $company = "Firma";

  public static $product = "Produkt";
```

And so on... Point is, that every email have this variables and you dont have to repeat it everywhere.

I made similar behavior for neon:
```
// default.cs_CZ.neon
message:
  hello: "Ahoj světe 2"
```

```
// main.cs_CZ.neon
includes:
  - default.cs_CZ.neon

message:
  hello: "Ahoj světe"
```

It inserts variables to the array and you are able to overwrite them if you need it.

What do you think about this use case? Is it acceptable for you?

**Comment by Olicek** [comment]
What do you thing about this idea? Is it bad idea? Have you better solution? I will need something similar soon. My question is if i should make some adapter for this feature, prepare PR or wait for your solutin? :-)

**Comment by fprochazka** [comment]
I do like the `includes` idea. And it would probably be simple to implement.

**Comment by Olicek** [comment]
it is. i had it, but i lost it :D I can prepare it again, if you like.

**Comment by fprochazka** [comment]
Ideally, it should use the config loader from Nette, it already implements the `includes` behavior. If you have the time a PR would be nice :)
